tools/lambda-promtail/lambda-promtail/s3.go

```diff
@@ -18,24 +20,65 @@ import (
-
-	// regex that extracts the timestamp (RFC3339) from message log
-	timestampRegex = regexp.MustCompile(`\w+ (?P<timestamp>\d+-\d+-\d+T\d+:\d+:\d+\.\d+Z)`)
-)
+type logDefinition struct {
+	logType        string
+	timestampType  string
+	filenameRegex  *regexp.Regexp
+	timestampRegex *regexp.Regexp
+	skipHeader     bool
+}
 
 const (
-	FLOW_LOG_TYPE string = "vpcflowlogs"
-	LB_LOG_TYPE   string = "elasticloadbalancing"
+	S3_LB       string = "s3_lb"
+	S3_VPC_FLOW string = "s3_vpc_flow"
+	S3_WAF             = "s3_waf"
+)
+
+var (
+	logDefinitionList = []logDefinition{
+		logDefinition{
+			logType:       S3_LB,
+			timestampType: "rfc3339",
+			// regex that parses the log file name fields for AWS Loadbalancer logs in S3
+			// source:  https://docs.aws.amazon.com/elasticloadbalancing/latest/application/load-balancer-access-logs.html#access-log-file-format
+			// format:  bucket[/prefix]/AWSLogs/aws-account-id/elasticloadbalancing/region/yyyy/mm/dd/aws-account-id_elasticloadbalancing_region_app.load-balancer-id_end-time_ip-address_random-string.log.gz
+			// example: my-bucket/AWSLogs/123456789012/elasticloadbalancing/us-east-1/2022/01/24/123456789012_elasticloadbalancing_us-east-1_app.my-loadbalancer.b13ea9d19f16d015_20220124T0000Z_0.0.0.0_2et2e1mx.log.gz
+			filenameRegex: regexp.MustCompile(`AWSLogs\/(?P<account_id>\d+)\/elasticloadbalancing\/(?P<region>[\w-]+)\/(?P<year>\d+)\/(?P<month>\d+)\/(?P<day>\d+)\/\d+\_elasticloadbalancing\_\w+-\w+-\d_(?:(?:app|nlb|net)\.*?)?(?P<src>[a-zA-Z0-9\-]+)`),
+			// regex that extracts the timestamp from the AWS Loadbalancer message log
+			// format: timestamp (RFC3339)
+			// example: h2 2022-12-20T23:55:02.599911Z ...
+			timestampRegex: regexp.MustCompile(`\w+ (?P<timestamp>\d+-\d+-\d+T\d+:\d+:\d+\.\d+Z)`),
+			skipHeader:     false,
+		},
+		logDefinition{
+			logType:       S3_VPC_FLOW,
+			timestampType: "seconds",
+			// regex that parses the log file name fields for AWS VPC Flow Logs in S3
+			// source: https://docs.aws.amazon.com/vpc/latest/userguide/flow-logs-s3.html#flow-logs-s3-path
+			// format: bucket-and-optional-prefix/AWSLogs/account_id/vpcflowlogs/region/year/month/day/aws_account_id_vpcflowlogs_region_flow_log_id_YYYYMMDDTHHmmZ_hash.log.gz
+			// example: 123456789012_vpcflowlogs_us-east-1_fl-1234abcd_20180620T1620Z_fe123456.log.gz
+			filenameRegex: regexp.MustCompile(`AWSLogs\/(?P<account_id>\d+)\/vpcflowlogs\/(?P<region>[\w-]+)\/(?P<year>\d+)\/(?P<month>\d+)\/(?P<day>\d+)\/\d+\_vpcflowlogs\_\w+-\w+-\d_(?:(?:app|nlb|net)\.*?)?(?P<src>[a-zA-Z0-9\-]+)`),
+			// regex that extracts the timestamp from the AWS VPC Flow Logs message log
+			// format: timestamp (seconds)
+			// example: ... 1669842701 1669842702 ACCEPT
+			timestampRegex: regexp.MustCompile(`(?P<begin_date>\d+) (?P<end_date>\d+) (?:ACCEPT|REJECT)`),
```

Comment:
FWIW I explicitly ignored them as they weren't loki standard and I felt that more advanced parsing should be left to actual promtail rather than this lambda becoming a full parser too